<h1>Data Analytics Project</h1>

<h1>United Kingdom Road Accident Data Analysis</h1>
<h3>Inclusive Years : 2019-2020</h3>
<h5>Analyst: Cristina Gail Rodelas</h5>

In [96]:
import numpy as np
import pandas as pd
import warnings
#import matplotlib.pyplot as plt
#import seaborn as sns
from scipy.stats import f_oneway
warnings.filterwarnings('ignore')

In [97]:
accident = pd.read_csv('dataset\\accident_data.csv')

In [98]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26/08/2019  51.517715   
3       200701BS66128           Serious    16/08/2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18/02/2022  57.374005   
660675  201091NM01881            Slight    21/02/2022  57.232273   
660676  201091NM01935            Slight    23/02/2022  57.585044   
660677  201091NM01964           Serious    23/02/2022  57.214898   
660678  201091NM02142           Serious    28/02/2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<h1>Data Preparation</h1>

In [99]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [100]:
accident.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

In [101]:
accident.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [102]:
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')

In [103]:
accident['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

In [104]:
accident.dtypes

Index                        object
Accident_Severity          category
Accident Date                object
Latitude                    float64
Light_Conditions             object
District Area                object
Longitude                   float64
Number_of_Casualties          int64
Number_of_Vehicles            int64
Road_Surface_Conditions      object
Road_Type                    object
Urban_or_Rural_Area          object
Weather_Conditions           object
Vehicle_Type                 object
dtype: object

<h1>Data Cleansing</h1>
<h3>Filling out null values</h3>

In [105]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [106]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mode()[0])
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mode()[0])  # Corrected here
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna('unknown surface condition')
accident['Road_Type'] = accident['Road_Type'].fillna('unaccounted')
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna('unaccounted')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])

In [107]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

<h1>Categorical Data Fields</h1>

In [108]:
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')
accident['Latitude'] = accident['Latitude'].astype('category')
accident['Light_Conditions'] = accident['Light_Conditions'].astype('category')
accident['District Area'] = accident['District Area'].astype('category')
accident['Longitude'] = accident['Longitude'].astype('category')
#accident['Number_of_Casualties'] = accident['Number_of_Casualties'].astype('int')
#accident['Number_of_Vehicles'] = accident['Number_of_Vehicles'].astype('int')
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category')
accident['Road_Type'] = accident['Road_Type'].astype('category')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].astype('category')
accident['Weather_Conditions'] = accident['Weather_Conditions'].astype('category')
accident['Vehicle_Type'] = accident['Vehicle_Type'].astype('category')

In [109]:
accident.dtypes

Index                        object
Accident_Severity          category
Accident Date                object
Latitude                   category
Light_Conditions           category
District Area              category
Longitude                  category
Number_of_Casualties          int64
Number_of_Vehicles            int64
Road_Surface_Conditions    category
Road_Type                  category
Urban_or_Rural_Area        category
Weather_Conditions         category
Vehicle_Type               category
dtype: object

<h1>Converting object to DateTime Data Type</h1>

In [110]:
accident['Accident Date'] = accident['Accident Date'].astype('str')
#remove any space
accident['Accident Date'] = accident['Accident Date'].str.strip()
#convert everything from / (slash) to - (dash)
accident['Accident Date'] = accident['Accident Date'].str.replace('/', '-')

In [111]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst = True, errors = 'coerce')

<h1>Adjusting Data Types</h1>

In [112]:
#accident['Accident Date']
accident.dtypes

Index                              object
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                         category
Light_Conditions                 category
District Area                    category
Longitude                        category
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
dtype: object

<h1>Extracting date information using pandas date time</h1>

In [113]:
accident['Year'] = accident['Accident Date'].dt.year
accident['Month'] = accident['Accident Date'].dt.month
accident['Day'] = accident['Accident Date'].dt.day
accident['DayofWeek'] = accident['Accident Date'].dt.dayofweek #Monday=0, Sunday=6

In [114]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
Year                       0
Month                      0
Day                        0
DayofWeek                  0
dtype: int64

<h1>Data Analytics</h1>
<h3>Analyzing Each from the Data set</h3>

<h1>UniVariate</h1>

In [115]:
df_2019 = accident[accident['Year'] == 2019]

<h1>Univariate Analysis from Year 2019</h1>
<h1>Question No.1: How many incidents happened under light conditions?</h1>

In [116]:
print(f"total number of accident in 2019 is {df_2019['Light_Conditions'].value_counts().sum()}")
df_2019['Light_Conditions'].value_counts()

total number of accident in 2019 is 182115


Light_Conditions
Daylight                       133042
Darkness - lights lit           35769
Darkness - no lighting          10842
Darkness - lighting unknown      1768
Darkness - lights unlit           694
Name: count, dtype: int64

<h3>Insight No.1</h3>
<h4>Contrary to popular belief the most accident happens in darkness or at night time, majority of the accident happens on broad daylight. About 73.05% of the total accident in 2019 happened during Daylight.</h4>

<h1>Question No.2: What is the total number of accidents in 2019 and 2020?</h1>

In [117]:
total_accidents_2019 = accident[accident['Accident Date'].dt.year == 2019]['Accident Date'].count()
total_accidents_2020 = accident[accident['Accident Date'].dt.year == 2020]['Accident Date'].count()
total_accidents = total_accidents_2019 + total_accidents_2020
total_accidents

352706

<h3>Insight No.2</h3>
<h4>There are 352,706 accidents recorded across 2019 and 2020.</h4>

<h1>Question No.3: What is the average number of accidents per month?</h1>

In [118]:
average_accidents_per_month = accident['Accident Date'].groupby(accident['Accident Date'].dt.month).count().mean()
average_accidents_per_month

55056.583333333336

<h3>Insight No.3</h3>
<h4>The average number of accidents per month is 55,056.58.</h4>

<h1>Question No.4: What is the most frequent accident severity?</h1>

In [119]:
severity_counts = accident['Accident_Severity'].value_counts()
severity_counts

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<h3>Insight No.4</h3>
<h4>Most accidents are classified as Slight (563,801 accidents), followed by Serious (88,217 accidents) and Fatal (8,661 accidents).</h4>

<h1>Question No.5: What is the most common light condition during accidents?</h1>

In [120]:
most_common_light_condition = accident['Light_Conditions'].value_counts()
most_common_light_condition

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<h3>Insight No.5</h3>
<h4>The highest number of accidents happened in Daylight, with 484,880 recorded incidents. Meanwhile, 129,335 accidents occurred in Darkness with street lights lit.</h4>

<h1>Question No.6: What is the most frequent weather condition during accidents?</h1>

In [121]:
most_frequent_weather = accident['Weather_Conditions'].value_counts()
most_frequent_weather

Weather_Conditions
Fine no high winds       520885
Raining no high winds     79696
Other                     17150
unaccounted               14128
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h3>Insight No.6</h3>
<h4>The most common weather condition during accidents is "Fine no high winds" with 520,885 recorded incidents. </h4>

<h1>Question No.7: What is the average number of vehicles involved in each accident?</h1>

In [122]:
average_vehicles = accident['Number_of_Vehicles'].mean()
average_vehicles

1.8312554205597575

<h3>Insight No.7</h3>
<h4>Each accident involves an average of 1.83 vehicles.</h4>

<h1>Question No.8: What is the most common road surface condition where accidents occurred?</h1>

In [123]:
most_common_surface = accident['Road_Surface_Conditions'].value_counts()
most_common_surface

Road_Surface_Conditions
Dry                          447821
Wet or damp                  186708
Frost or ice                  18517
Snow                           5890
Flood over 3cm. deep           1017
unknown surface condition       726
Name: count, dtype: int64

<h3>Insight No.8</h3>
<h4>The majority of accidents occurred on dry roads, with 447,821 recorded incidents.</h4>

<h1>Question No.9: Which day of the week had the highest number of accidents?</h1>

In [124]:
day_accidents = accident['DayofWeek'].value_counts()
day_accidents

DayofWeek
5    107178
2     99558
3     99511
4     97900
1     94550
6     89302
0     72680
Name: count, dtype: int64

<h3>Insight No.9</h3>
<h4>Friday (Day 5) recorded the highest number of accidents, with 107,178 incidents.</h4>

<h1>Question No.10: Which road type experienced the most accidents?</h1>

In [125]:
road_type_counts = accident['Road_Type'].value_counts()
road_type_counts

Road_Type
Single carriageway    492143
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
unaccounted             4520
Name: count, dtype: int64

<h3>Insight No.10</h3>
<h4>The most accidents occurred on Single Carriageway roads, with 492,143 recorded incidents. </h4>

<h1>BiVariate</h1>

<h1>Question No.11: How does weather affect the number of accidents?</h1>

In [126]:
weather_accidents = accident.groupby('Weather_Conditions')['Accident Date'].count().sort_values(ascending=False)
weather_accidents

Weather_Conditions
Fine no high winds       520885
Raining no high winds     79696
Other                     17150
unaccounted               14128
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: Accident Date, dtype: int64

<h3>Insight No.11</h3>
<h4>The highest number of accidents occurred under fine weather with no high winds (520,885 cases), likely due to more vehicles being on the road. However, rainy conditions (79,696 cases) and fog/mist (3,528 cases) significantly increased accident rates compared to normal weather.</h4>

<h1>Question No.12: What is the average number of casualties on different road surface conditions?</h1>

In [127]:
avg_casualties_per_surface = accident.groupby('Road_Surface_Conditions')['Number_of_Casualties'].mean()
avg_casualties_per_surface

Road_Surface_Conditions
Dry                          1.334736
Flood over 3cm. deep         1.495575
Frost or ice                 1.348815
Snow                         1.359593
Wet or damp                  1.411064
unknown surface condition    1.216253
Name: Number_of_Casualties, dtype: float64

<h3>Insight No.12</h3>
<h4>It shows that roads with flooding over 3cm deep have the highest average number of casualties at 1.50 per accident, followed by wet or damp roads at 1.41 casualties per accident. Roads covered in snow (1.36), frost or ice (1.35), and dry conditions (1.33) also present risks. While, the lowest casualty rate is observed in unknown surface conditions (1.22 casualties per accident).</h4>

<h1>Question No.13: How do weather conditions relate to light conditions in road accidents?</h1>

In [128]:
weather_light_condition = accident.groupby(['Weather_Conditions', 'Light_Conditions'])['Accident Date'].count()
weather_light_condition

Weather_Conditions     Light_Conditions           
Fine + high winds      Darkness - lighting unknown        59
                       Darkness - lights lit            1852
                       Darkness - lights unlit            51
                       Darkness - no lighting            796
                       Daylight                         5796
Fine no high winds     Darkness - lighting unknown      3683
                       Darkness - lights lit           91951
                       Darkness - lights unlit          1700
                       Darkness - no lighting          24867
                       Daylight                       398684
Fog or mist            Darkness - lighting unknown        65
                       Darkness - lights lit             921
                       Darkness - lights unlit            37
                       Darkness - no lighting            862
                       Daylight                         1643
Other                  Darkness - 

<h3>Insight No.13</h3>
<h4>Most accidents occur in clear weather during daylight, but foggy and rainy conditions increase accidents in low-light conditions. </h4>

<h1>MultiVariate</h1>

In [129]:
#Accident Per Year
accidents_per_year = accident['Year'].value_counts().sort_index()

#Float might be caused of values without year
# accident = accident['Year'].dropna()
# Accident per month
accidents_per_month = accident['Month'].value_counts().sort_index()

In [130]:
accidents_per_year

Year
2019    182115
2020    170591
2021    163554
2022    144419
Name: count, dtype: int64

In [131]:
accidents_per_month

Month
1     52872
2     49491
3     54086
4     51744
5     56352
6     56481
7     57445
8     53913
9     56455
10    59580
11    60424
12    51836
Name: count, dtype: int64

<h1>Question No.14: How does bad weather combined with poor road surface conditions affect accident risks?</h1>

In [132]:
bad_weather_accidents = accident[(accident['Weather_Conditions'] == 'Rain') | 
                           (accident['Weather_Conditions'] == 'Snow') | 
                           (accident['Weather_Conditions'] == 'Fog') | 
                           (accident['Weather_Conditions'] == 'Storm')]

poor_road_accidents = accident[(accident['Road_Surface_Conditions'] == 'Wet or damp') | 
                         (accident['Road_Surface_Conditions'] == 'Snow') | 
                         (accident['Road_Surface_Conditions'] == 'Frost or ice') | 
                         (accident['Road_Surface_Conditions'] == 'Flood over 3cm')]

combined_accidents = bad_weather_accidents['Accident Date'].count() + poor_road_accidents['Accident Date'].count()

combined_accidents

211115

<h3>Insight No.14</h3>
<h4>A total of 211,115 accidents occurred under bad weather conditions (rain, snow, fog, or storms) combined with poor road surfaces (wet, icy, snowy, or flooded roads). </h4>

<h1>Question No.15: Do winter months have higher accident rates due to unwanted weather conditions?</h1>

In [133]:
winter_accidents = accident[(accident['Month'] == 12) | (accident['Month'] == 1) | (accident['Month'] == 2)]['Month'].count()
total_accidents = accident['Month'].count()

winter_percentage = (winter_accidents / total_accidents) * 100
winter_accidents, winter_percentage

(154199, 23.339473481070232)

<h3>Insight No.15</h3>
<h4>It shows that 154,199 accidents occurred during the winter months, making up 23.34% of the total accidents.</h4>

<h1>Question No.16: How do weather conditions, road surface conditions, and accident severity interact?</h1>

In [134]:
weather_road_severity = accident.groupby(['Weather_Conditions', 'Road_Surface_Conditions'])['Accident_Severity'].value_counts()
weather_road_severity

Weather_Conditions  Road_Surface_Conditions    Accident_Severity
Fine + high winds   Dry                        Slight               4162
                                               Serious               758
                                               Fatal                 103
                    Flood over 3cm. deep       Slight                 14
                                               Fatal                   1
                                                                    ... 
unaccounted         Wet or damp                Serious               169
                                               Fatal                  29
                    unknown surface condition  Slight                550
                                               Serious                61
                                               Fatal                   1
Name: count, Length: 162, dtype: int64

<h3>Insight No.16</h3>
<h4>It shows that most accidents occur in "Fine + high winds" conditions on dry roads, with 4,162 slight accidents, 758 serious accidents, and 103 fatal accidents. However, accidents on flooded roads, though less frequent, show a higher fatality percentage.</h4>

<h1>Question No.17: How do road type, light conditions, and accident severity relate to each other?</h1>

In [135]:
road_light_severity = accident.groupby(['Road_Type', 'Light_Conditions'])['Accident_Severity'].value_counts()
road_light_severity

Road_Type         Light_Conditions             Accident_Severity
Dual carriageway  Darkness - lighting unknown  Slight                 805
                                               Serious                 93
                                               Fatal                    7
                  Darkness - lights lit        Slight               17025
                                               Serious               2762
                                                                    ...  
unaccounted       Darkness - no lighting       Serious                 33
                                               Fatal                    1
                  Daylight                     Slight                3098
                                               Serious                349
                                               Fatal                   22
Name: count, Length: 90, dtype: int64

<h3>Insight No.17</h3>
<h4>Most accidents happen on dual carriageways under daylight conditions, with 17,025 slight, 2,762 serious, and 22 fatal accidents.</h4>

<h1>Question No.18: How do vehicle type, urban/rural area, and accident severity correlate?</h1>

In [136]:
vehicle_area_severity = accident.groupby(['Vehicle_Type', 'Urban_or_Rural_Area'])['Accident_Severity'].value_counts()
vehicle_area_severity

Vehicle_Type                         Urban_or_Rural_Area  Accident_Severity
Agricultural vehicle                 Rural                Slight                 540
                                                          Serious                121
                                                          Fatal                   14
                                     Unallocated          Fatal                    0
                                                          Serious                  0
                                                                               ...  
Van / Goods 3.5 tonnes mgw or under  Unallocated          Slight                   0
                                                          Fatal                    0
                                     Urban                Slight               19135
                                                          Serious               2707
                                                          Fatal           

<h3>Insight No.18</h3>
<h4>Most accidents involving vans and goods vehicles happen in urban areas, with 19,135 slight, 2,707 serious, and 161 fatal accidents. </h4>

<h1>Question No.19: How does the combination of weather conditions, road type, and accident severity affect the number of casualties in road accidents across different years?</h1>

In [137]:
# Grouping by year, weather conditions, road type, and accident severity
casualties_by_factors = accident.groupby(['Year', 'Weather_Conditions', 'Road_Type', 'Accident_Severity'])['Number_of_Casualties'].sum().reset_index()
casualties = casualties_by_factors.pivot_table(index=['Year', 'Weather_Conditions', 'Road_Type'],
                                                     columns='Accident_Severity',
                                                     values='Number_of_Casualties',
                                                     aggfunc='sum').fillna(0)

casualties

Accident_Severity                           Fatal  Serious  Slight
Year Weather_Conditions Road_Type                                 
2019 Fine + high winds  Dual carriageway       22       67     548
                        One way street          2        9      45
                        Roundabout              1       19     178
                        Single carriageway     92      415    2004
                        Slip road               0        4      35
...                                           ...      ...     ...
2022 unaccounted        One way street          0        7      67
                        Roundabout              1       11     303
                        Single carriageway     29      233    2130
                        Slip road               0        0      41
                        unaccounted             0        3      82

[216 rows x 3 columns]

<h3>Insight No.19</h3>
<h4>Accidents in 2019 and 2022 show that dual carriageways and single carriageways are the most common road types for accidents, with serious accidents occurring more often on dual carriageways.</h4>

<h1>Question No.20: How do different vehicle types and road conditions contribute to the severity of road accidents?</h1>

In [138]:
accident_severity_by_factors = accident.groupby(['Vehicle_Type', 'Road_Surface_Conditions', 'Accident_Severity']).size().unstack().fillna(0)
accident_severity_by_factors_sorted = accident_severity_by_factors.sort_values(by=['Fatal', 'Serious'], ascending=False)

accident_severity_by_factors_sorted

Accident_Severity                                               Fatal  \
Vehicle_Type                         Road_Surface_Conditions            
Car                                  Dry                         4403   
                                     Wet or damp                 1981   
Van / Goods 3.5 tonnes mgw or under  Dry                          311   
Bus or coach (17 or more pass seats) Dry                          220   
Motorcycle over 500cc                Dry                          217   
...                                                               ...   
Ridden horse                         Flood over 3cm. deep           0   
                                     Frost or ice                   0   
                                     Snow                           0   
                                     Wet or damp                    0   
                                     unknown surface condition      0   

Accident_Severity                                               Serious  \
Vehicle_Type                         Road_Surface_Conditions              
Car                                  Dry                          46458   
                                     Wet or damp                  17905   
Van / Goods 3.5 tonnes mgw or under  Dry                           3165   
Bus or coach (17 or more pass seats) Dry                           2367   
Motorcycle over 500cc                Dry                           2412   
...                                                                 ...   
Ridden horse                         Flood over 3cm. deep             0   
                                     Frost or ice                     0   
                                     Snow                             0   
                                     Wet or damp                      0   
                                     unknown surface condition        0   

Accident_Severity                                               Slight  
Vehicle_Type                         Road_Surface_Conditions            
Car                                  Dry                        286450  
                                     Wet or damp                120878  
Van / Goods 3.5 tonnes mgw or under  Dry                         19606  
Bus or coach (17 or more pass seats) Dry                         15017  
Motorcycle over 500cc                Dry                         14809  
...                                                                ...  
Ridden horse                         Flood over 3cm. deep            0  
                                     Frost or ice                    0  
                                     Snow                            0  
                                     Wet or damp                     1  
                                     unknown surface condition       0  

[96 rows x 3 columns]

<h3>Insight No.20</h3>
<h4>Most accidents, especially fatal ones, occur on dry roads, with cars having the highest frequency. Wet or damp roads also contribute to accidents, but they are less severe compared to dry conditions, especially for vehicles like buses and motorcycles.</h4>

<h1>Question 21</h1>
<h1>Which district areas have the highest accident frequency for the differenet levels of accident severity?</h1>

In [139]:
accident_by_area_severity = accident.groupby(['District Area', 'Accident_Severity']).size().reset_index(name='Accident_Count')
accident_by_area_severity 

District Area Accident_Severity  Accident_Count
0     Aberdeen City             Fatal              12
1     Aberdeen City           Serious             239
2     Aberdeen City            Slight            1072
3     Aberdeenshire             Fatal              66
4     Aberdeenshire           Serious             463
...             ...               ...             ...
1261    Wyre Forest           Serious             132
1262    Wyre Forest            Slight             815
1263           York             Fatal              21
1264           York           Serious             255
1265           York            Slight            1621

[1266 rows x 3 columns]

<h1>Insight 21</h1>
<h3>The district in Aberdeenshire with the highest accident frequency also has the greatest number of fatal accidents. Meanwhile, urban areas like York experience fewer, less severe accidents.</h3>

<h1>Question 22</h1>
<h1>Which month has the highest number of accidents based on the dataset?</h1>

In [140]:
accidents_by_month = accident.groupby('Month').size()
most_accidents = accidents_by_month.sort_values(ascending=False).head(1)

most_accidents

Month
11    60424
dtype: int64

<h1>Insight 22</h1>
<h3>The month with the highest number of accidents is November, with 60,424 accidents.</h3>

<h1>Question 23</h1>
<h1>How many accidents occurred in rural and urban areas in each district?</h1>

In [141]:
accidents_by_area = accident.groupby(['District Area', 'Urban_or_Rural_Area']).size().unstack(fill_value=0)
accidents_by_area

Urban_or_Rural_Area  Rural  Unallocated  Urban
District Area                                 
Aberdeen City          207            0   1116
Aberdeenshire         1781            0    149
Adur                   197            0    422
Allerdale              880            0    248
Alnwick                232            0      0
...                    ...          ...    ...
Wychavon              1088            0    273
Wycombe                739            0    990
Wyre                   480            0    758
Wyre Forest            419            0    550
York                   706            0   1191

[422 rows x 3 columns]

<h1>Insight 23</h1>
<h3>In most districts, there are more accidents in urban areas than in rural areas, although some districts like Aberdeenshire have more accidents in rural areas.</h3>